### PR TITLE
Add NI-SCOPE

### LINKED ISSUE #182: Add NI-SCOPE

### PR DESCRIPTION
**Pull request by texasaggie97**
[X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).
[X] I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.
~~[ ] I've added any unit tests that are applicable for this pull request~~
### What does this Pull Request accomplish?
* Adds NI-SCOPE
  * Metadata
  * Build
  * Placeholder example
  * Simple system test
### List issues fixed by this Pull Request below, if any.
* Fix #182 

### What testing has been done?
* Unit
* System

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 
-ALL_DRIVERS := nifake nifgen nidcpower nidmm niswitch nimodinst
+ALL_DRIVERS := nifake niscope nifgen nidcpower nidmm niswitch nimodinst
```
Comment (marcoskirsch) [inline Makefile]: While the ordering here is not important, it appears random and also inconsistent throughout the codebase whenever we list the drivers.
How about we fix that?
(could be separate PR if you think it's cleaner)

File: README.rst [hunk 2]
```diff
@@ -10,6 +10,10 @@
 |                      | NI-ModInst | |ModInstLatestVersion| |ModInstPythonVersion| |
 |                      +------------+-----------------------------------------------+
 |                      | NI-SWITCH  | |SwitchLatestVersion| |SwitchPythonVersion|   |
+|                      +------------+-----------------------------------------------+
+|                      | NI-FGEN    | |FgenLatestVersion| |FgenPythonVersion|       |
```
Comment (marcoskirsch) [inline README.rst]: another cleanup task: besides fixing order, perhaps we should clear all these variables (FgenPythonVersion, ModInstLatestVersion) and use the module names exactly as prefixes (nimodinstLatestVersion). Anything else seems arbitrary.

File: build/templates/session.py.mako [hunk 3]
```diff
@@ -97,13 +97,19 @@ You can specify a subset of repeated capabilities using the Python index notatio
...
+
+# Handle attribute names that start with a number
+beginning_underscore = ''
+if attributes[attribute]['name'][0].isdigit():
+    beginning_underscore = '_'
+
 %>\
     %if attributes[attribute]['enum']:
-    ${attributes[attribute]['name'].lower()} = attributes.AttributeEnum(attributes.Attribute${attributes[attribute]['type']}, enums.${attributes[attribute]['enum']}, ${attribute})
+    ${beginning_underscore + attributes[attribute]['name'].lower()} = attributes.AttributeEnum(attributes.Attribute${attributes[attribute]['type']}, enums.${attributes[attribute]['enum']}, ${attribute})
```
Comment (marcoskirsch) [inline build/templates/session.py.mako]: Is the python_name for the attribute not in metadata?
That seems weirdly inconsistent with function.

Assuming I understand correctly, this should be moved to "metadata expansion" phase.

File: docs/niscope/attributes.rst [hunk 4]
```diff
@@ -0,0 +1,6556 @@
+niscope.Session properties
+==========================
+
+.. py:currentmodule:: niscope
+
+.. py:attribute:: 5102_adjust_pretrigger_samples
```
Comment (marcoskirsch) [inline docs/niscope/attributes.rst]: I imagine this is missing a leading underscore?

File: src/niscope/metadata/functions_addon.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+# These dictionaries are applied to the generated functions dictionary at build time
+# Any changes to the API should be made here. functions.py is code generated
+
+# By default all functions in functions.py will be generated as a public function
+# This will override that with private - add '_' to the beginning of the name, or
+# don't generate at all
+functions_codegen_method = {
+    'InitWithOptions':                 { 'codegen_method': 'private',  },
+    'Initiate':                        { 'codegen_method': 'private',  },
```
Comment (marcoskirsch) [inline src/niscope/metadata/functions_addon.py]: InitiateAcquisition

File: Makefile (hunk 1 above)
Comment (texasaggie97) [reply Makefile]: What is the preferred order? Alphabetical other than nifake first?

Comment (marcoskirsch) [reply Makefile]: I think order will be arbitrary no matter what. Let's just keep it consistent.
Maybe:

1. NI-FAKE
2. "Normal" drivers in alphabetical order.
3. NI-ModInst

?

Comment (texasaggie97) [reply Makefile]: Changed

File: README.rst (hunk 2 above)
Comment (texasaggie97) [reply README.rst]: Changed

File: build/templates/session.py.mako (hunk 3 above)
Comment (texasaggie97) [reply build/templates/session.py.mako]: Moved to add_all

File: src/niscope/metadata/functions_addon.py (hunk 5 above)
Comment (texasaggie97) [reply src/niscope/metadata/functions_addon.py]: Fixed

File: src/niscope/metadata/functions_addon.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
+# TODO(texasaggie97) update this file with all NI-FGEN function information
```
Comment (marcoskirsch) [inline src/niscope/metadata/functions_addon.py]: NI-SCOPE?
You can reference #519 or remove altogether since the issue exists.

File: src/niscope/metadata/functions_addon.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    'GetNextInterchangeWarning':       { 'codegen_method': 'no',       },  # Not applicable to Python API
+    'ResetInterchangeCheck':           { 'codegen_method': 'no',       },  # Not applicable to Python API
+    'ClearInterchangeWarnings':        { 'codegen_method': 'no',       },  # Not applicable to Python API
+    'GetNextCoercionRecord':           { 'codegen_method': 'no',       },  # Not applicable to Python API
+    'error_query':                     { 'codegen_method': 'no',       },
+    'revision_query':                  { 'codegen_method': 'no',       },
+    'SampleMode':                      { 'codegen_method': 'no',       },  # Equivalent attribute is available
+    'GetNormalizationCoefficients':    { 'codegen_method': 'no',       },  # Has void param
+    'GetScalingCoefficients':          { 'codegen_method': 'no',       },  # Has void param
+    'Fetch':                           { 'codegen_method': 'no',       },  # Has niScope_wfmInfo param
```
Comment (marcoskirsch) [inline src/niscope/metadata/functions_addon.py]: point to an issue for handling this to make clear it's temporary

File: src/niscope/metadata/functions_addon.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+#                       interpreting the return value as a size rather than an error.
+#                       'value' should be the name of the parameter through which the size (0, then the real
+#                       one) is passed in. This parameter won't exist in the corresponding Python Session method.
+functions_buffer_info = {
+    'GetError':                     { 'parameters': { 3: { 'size': {'mechanism':'ivi-dance', 'value':'bufferSize'}, }, }, },
+    'self_test':                    { 'parameters': { 2: { 'size': {'mechanism':'fixed', 'value':256}, }, }, }, # From documentation
+    'GetAttributeViString':         { 'parameters': { 4: { 'size': {'mechanism':'ivi-dance', 'value':'bufSize'}, }, }, },
+    'GetCalUserDefinedInfo':        { 'parameters': { 1: { 'size': {'mechanism':'fixed', 'value':256}, }, }, }, # From LabVIEW VI, even though niDMM_GetCalUserDefinedInfoMaxSize() exists.
+    'error_message':                { 'parameters': { 2: { 'size': {'mechanism':'fixed', 'value':256}, }, }, }, # From documentation
+    'init':                         { 'parameters': { 0: { 'is_buffer': True, }, }, },
```
Comment (marcoskirsch) [inline src/niscope/metadata/functions_addon.py]: 'is_buffer': True should not be necessary anymore.

File: tox.ini [hunk 9]
```diff
@@ -21,6 +21,7 @@ description =
     nidcpower_system_tests: Run system tests for nidcpower
     nidmm_system_tests: Run system tests for nidmm
     niswitch_system_tests: Run system tests for niswitch
+    niscope_system_tests: Run system tests for niswitch
```
Comment (marcoskirsch) [inline tox.ini]: for niscope

File: src/niscope/metadata/functions_addon.py (hunk 6 above)
Comment (texasaggie97) [reply src/niscope/metadata/functions_addon.py]: Done

File: src/niscope/metadata/functions_addon.py (hunk 7 above)
Comment (texasaggie97) [reply src/niscope/metadata/functions_addon.py]: Added

File: src/niscope/metadata/functions_addon.py (hunk 8 above)
Comment (texasaggie97) [reply src/niscope/metadata/functions_addon.py]: Removed

File: tox.ini (hunk 9 above)
Comment (texasaggie97) [reply tox.ini]: Fixed
